src/main/java/uk/gov/pay/products/resources/ProductResource.java

```diff
@@ -88,6 +105,38 @@ public Response findProductByGatewayAccountIdAndExternalId(@PathParam("gatewayAc
                         Response.status(NOT_FOUND).build());
     }
 
+    @POST
+    @Path("/products/{productExternalId}/regenerate-api-key")
+    @Produces(APPLICATION_JSON)
+    @Consumes(APPLICATION_JSON)
+    public Response regenerateProductApiKey(@PathParam("productExternalId") String productExternalId) {
+        return productFactory.productFinder().findByExternalId(productExternalId).map((product) -> {
+            CreateTokenRequest createTokenRequest = new CreateTokenRequest(
+                    String.valueOf(product.getGatewayAccountId()),
+                    getDescription(product),
+                    emailAddress,
+                    CARD,
+                    PRODUCTS);
+
+            Response response = client.target(publicAuthUrl)
+                    .request()
+                    .post(Entity.entity(createTokenRequest, MediaType.APPLICATION_JSON));
+
+            TokenResponse tokenResponse = response.readEntity(TokenResponse.class);
+            String token;
+            if (tokenResponse != null && tokenResponse.getToken() != null && !tokenResponse.getToken().isEmpty()) {
```

Comment:
`Optional.ofNullable(tokenResponse).map(TokenResponse::getToken).filter(not(String::isEmpty)).map(…)` might make this is a little more readable.

But I question if we need the `isEmpty()` check? Do we really think public auth is likely to return an empty string?

Reply:
I don’t think we have to worry too much about intercepted responses between microservices given they’re all internal HTTPS calls. Even if products did store an empty string, it would mean the API calls it makes to public API would fail (because the API key wouldn’t be a working one).

Reply:
Thanks. I will drop `isEmpty()`.